subsys/jwt/jwt_psa.c

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright (C) 2024 BayLibre SAS
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <string.h>
+#include <zephyr/types.h>
+#include <errno.h>
+
+#include <zephyr/data/jwt.h>
+#include <zephyr/data/json.h>
+
+#include <psa/crypto.h>
+
+int jwt_sign_impl(struct jwt_builder *builder, const unsigned char *der_key, size_t der_key_len,
+		  unsigned char *sig, size_t sig_size)
+{
+	psa_status_t status;
+	psa_key_attributes_t attr = PSA_KEY_ATTRIBUTES_INIT;
+	psa_key_id_t key_id = 0;
+	size_t sig_len_out;
+	psa_algorithm_t alg;
+	int ret;
+
+#if defined(CONFIG_JWT_SIGN_ECDSA)
+	psa_set_key_type(&attr, PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_SECP_R1));
+	psa_set_key_algorithm(&attr, PSA_ALG_ECDSA(PSA_ALG_SHA_256));
+	alg = PSA_ALG_ECDSA(PSA_ALG_SHA_256);
+#else /* CONFIG_JWT_SIGN_RSA */
+	psa_set_key_type(&attr, PSA_KEY_TYPE_RSA_KEY_PAIR);
+	psa_set_key_algorithm(&attr, PSA_ALG_RSA_PKCS1V15_SIGN(PSA_ALG_SHA_256));
+	alg = PSA_ALG_RSA_PKCS1V15_SIGN(PSA_ALG_SHA_256);
+#endif /* CONFIG_JWT_SIGN_ECDSA || CONFIG_JWT_SIGN_RSA */
+	psa_set_key_usage_flags(&attr, PSA_KEY_USAGE_SIGN_MESSAGE);
+
+	status = psa_import_key(&attr, der_key, der_key_len, &key_id);
```

Comment:
No more setting of the key bits?

Reply:
Yes, it's not strictly needed. The key data being passed will automatically determine the key's length: https://github.com/zephyrproject-rtos/mbedtls/blob/fb36f3fe20f9f62e67b1a20c0cfe0a6788ec2bf6/include/psa/crypto.h#L586